### PR TITLE
Upgrade Node Import to Blender 4.x

### PR DESCRIPTION
**Pull request by Spiderguy-F**
The node system has changed slightly. These adjustments should cover it. Tested for creating nodes and exporting gltf including behavior graphs. But not all.

### CODE REVIEW

**Comment by Exairnous** [comment]
Untested, but this may help to support Blender versions 3.6 and below.
```
def draw(self, context, layout, node=None, text=None):

def draw_color(self, context, node=None):
```

**Comment by Spiderguy-F** [comment]
def draw(self, context, layout, node=None, text=None):
def draw_color(self, context, node=None):

I read up about overloads and it is not recommended in python. 
While the idea is good and it might just work, I realised, we also need the inheritance changed:
class BGHubsEntitySocketInterface(NodeSocket): (4)
class BGHubsEntitySocketInterface(NodeSocketInterface): (3)

So I think it's best to maintain 2 versions of the sockets.py and import the correct ones at the correct time. 
The sockets appear quite a lot in the code (nodes.py) and I noticed this: 
from .sockets import *

inside of def init() of 10 different classes (nodes.py) I found this: 
from .sockets import BGFlowSocket

rather than checking every time, can't we just use the BGFlowSocket class directly, since we already imported it (*) at the beginning of the script?

**Comment by Exairnous** [comment]
For future reference, we talked about this at the latest add-on/BG meeting and agreed that adding the keyword arguments to the draw functions and then doing an if check when importing the NodeSocket class and doing an import NodeSocketInterface as NodeSocket for Blender versions below 4.0 seems like it should work.

File: behavior_graph.py [hunk 1]
```diff
@@ -543,7 +542,7 @@ def gather_events_and_variables(export_settings):
 
 
 def gather_nodes(ob, ob_idx, slot, slot_idx, export_settings, events, variables, export_report):
-    from .sockets import BGFlowSocket
+    # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline behavior_graph.py:545]: We should remove this now that we import all the module

File: behavior_graph.py [hunk 2]
```diff
@@ -53,13 +53,13 @@ class BGTree(NodeTree):
     def mark_invalid_links(self):
         print("mark invalid links")
         for link in self.links:
-            if type(link.from_socket) != type(link.to_socket):
+            if type(link.from_socket) is not type(link.to_socket):
                 link.is_valid = False
 
     def update(self):
-        from .sockets import BGEnumSocket
+        # from .sockets import BGEnumSocket
```
Comment (keianhzo) [inline behavior_graph.py:60]: Same here, we don't nee to keep the commented lie

File: nodes.py [hunk 3]
```diff
@@ -34,15 +34,15 @@ def poll(cls, ntree):
 
 class BGEventNode():
     def init(self, context):
-        from .sockets import BGFlowSocket
+        # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline nodes.py:37]: Same

File: nodes.py [hunk 4]
```diff
@@ -277,7 +277,7 @@ class BGNode_media_onMediaEvent(BGEventNode, BGEntityPropertyNode, BGNode, Node)
     def init(self, context):
         self.use_custom_color = True
         self.color = (0.6, 0.2, 0.2)
-        from .sockets import BGFlowSocket
+        # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline nodes.py:280]: Same

File: nodes.py [hunk 5]
```diff
@@ -290,7 +290,7 @@ def draw_buttons(self, context, layout):
 
 
 def update_output_sockets(self, context):
-    from .sockets import BGFlowSocket
+    # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline nodes.py:293]: Same

File: nodes.py [hunk 6]
```diff
@@ -315,7 +315,7 @@ class BGNode_flow_sequence(BGNode, Node):
     def init(self, context):
         super().init(context)
         self.color = (0.2, 0.2, 0.2)
-        from .sockets import BGFlowSocket
+        # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline nodes.py:318]: Same

File: nodes.py [hunk 7]
```diff
@@ -1462,7 +1462,7 @@ class BGNode_media_mediaPlayback(BGNetworked, BGActionNode, BGNode, Node):
     def init(self, context):
         super().init(context)
         self.inputs.new("BGHubsEntitySocket", "entity")
-        from .sockets import BGFlowSocket
+        # from .sockets import BGFlowSocket
         BGFlowSocket.create(self.inputs, name="play")
```
Comment (keianhzo) [inline nodes.py:1466]: Same

File: nodes.py [hunk 8]
```diff
@@ -1479,7 +1479,7 @@ class BGNode_media_frame_setMediaFrameProperty(BGNetworked, BGNode, Node):
     def init(self, context):
         super().init(context)
         self.inputs.new("BGHubsEntitySocket", "entity")
-        from .sockets import BGFlowSocket
+        # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline nodes.py:1482]: Same

File: nodes.py [hunk 9]
```diff
@@ -1498,7 +1498,7 @@ def init(self, context):
         super().init(context)
         entity = self.inputs.new("BGHubsEntitySocket", "entity")
         entity.poll_components = "rigidbody"
-        from .sockets import BGFlowSocket
+        # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline nodes.py:1501]: Same

File: nodes.py [hunk 10]
```diff
@@ -1543,7 +1543,7 @@ class BGNode_animation_play(BGNetworked, BGActionNode, BGNode, Node):
 
     def init(self, context):
         super().init(context)
-        from .sockets import BGFlowSocket
+        # from .sockets import BGFlowSocket
```
Comment (keianhzo) [inline nodes.py:1546]: Same

File: utils.py [hunk 11]
```diff
@@ -133,7 +133,7 @@ def get_hubs_ext(export_settings):
     exts = export_settings["gltf_user_extensions"]
     for ext in exts:
         import io_hubs_addon
-        if type(ext) == io_hubs_addon.io.gltf_exporter.glTF2ExportUserExtension:
+        if type(ext) is io_hubs_addon.io.gltf_exporter.glTF2ExportUserExtension:
```
Comment (keianhzo) [inline utils.py:136]: We can also move the import outside here

File: sockets3.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
+#OBSOLETE, WILL BE REMOVED IN THE FUTURE
```
Comment (keianhzo) [inline sockets3.py:1]: I don't think we need this file anymore, maybe it was a leftover from testing?

**Review by keianhzo — CHANGES_REQUESTED** [review]
As I mentioned for the future it's good to separatte format updates from the actual functional updates in different PRs as it makes reviewing easier.

Codewise I only see a few nits and some code leftovers that should be easy to clean and fixing the merge conflicts.

I'll test with the demos.

**Comment by keianhzo** [comment]
I've seen some method signature updates that I've pushed on this PR: https://github.com/MozillaReality/blender-gltf-behavior-graph/pull/22

File: sockets3.py (hunk 12 above)
Comment (Spiderguy-F) [reply sockets3.py]: Yes, sockets3 became obsolete after out last brainstorm, I removed it. 
Also did more cleanup

**Comment by keianhzo** [comment]
Great, thanks. Now we just need to fix the conflicts so this can be merged

**Comment by keianhzo** [comment]
We still need to fix them, otherwise Github doesn't let you merge.

**Review by keianhzo — APPROVED** [review]

**Pull request merged**